internal/tofu/node_provider.go

```diff
@@ -26,33 +27,80 @@ var (
 )
 
 // GraphNodeExecutable
-func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) (diags tfdiags.Diagnostics) {
-	_, err := ctx.InitProvider(n.Addr)
-	diags = diags.Append(err)
+func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) tfdiags.Diagnostics {
+	instances, diags := n.initInstances(ctx, op)
+
+	for key, provider := range instances {
+		diags = diags.Append(n.executeInstance(ctx, op, key, provider))
+	}
+
+	return diags
+}
+func (n *NodeApplyableProvider) initInstances(ctx EvalContext, op walkOperation) (map[addrs.InstanceKey]providers.Interface, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	var initKeys []addrs.InstanceKey
+	if n.Config == nil || n.Config.Instances == nil || op == walkValidate {
+		initKeys = append(initKeys, addrs.NoKey)
+	} else {
+		// Instances are set AND we are not validating
+		for key := range n.Config.Instances {
+			initKeys = append(initKeys, key)
+		}
+	}
+
+	// init -> config (different due to validate skipping most for_each logic)
+	instanceKeys := make(map[addrs.InstanceKey]addrs.InstanceKey)
+	if n.Config == nil || n.Config.Instances == nil {
+		instanceKeys[addrs.NoKey] = addrs.NoKey
+	} else if op == walkValidate {
+		// Instances are set AND we are validating
+		for key := range n.Config.Instances {
+			instanceKeys[addrs.NoKey] = key
+		}
+	} else {
+		// Instances are set AND we are not validating
+		for key := range n.Config.Instances {
+			instanceKeys[key] = key
+		}
+	}
+
+	for _, key := range initKeys {
+		_, err := ctx.InitProvider(n.Addr, key)
+		diags = diags.Append(err)
+	}
 	if diags.HasErrors() {
-		return diags
+		return nil, diags
+	}
+
+	instances := make(map[addrs.InstanceKey]providers.Interface)
+	for initKey, configKey := range instanceKeys {
+		provider, _, err := getProvider(ctx, n.Addr, initKey)
+		diags = diags.Append(err)
+		instances[configKey] = provider
 	}
-	provider, _, err := getProvider(ctx, n.Addr)
-	diags = diags.Append(err)
 	if diags.HasErrors() {
-		return diags
+		return nil, diags
 	}
 
+	return instances, diags
+}
+func (n *NodeApplyableProvider) executeInstance(ctx EvalContext, op walkOperation, providerKey addrs.InstanceKey, provider providers.Interface) (diags tfdiags.Diagnostics) {
```

Comment:
super nit: we have a provider and then a key in other functions, let's keep the order consistent.

```suggestion
func (n *NodeApplyableProvider) executeInstance(ctx EvalContext, op walkOperation, provider providers.Interface, providerKey addrs.InstanceKey) (diags tfdiags.Diagnostics) {
```

Reply:
7fd09e5d2a